snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -270,25 +270,13 @@ def submit(self, message: Message[Tick]) -> None:
         # Periodically record the legnth of each buffer
         self.__record_tick_buffer_length()
 
-        # If there are any empty buffers, we can't submit anything yet.
-        # Otherwise if all the buffers have ticks then we look for the partition/s
-        # with the earliest tick (i.e. the tick with the earliest upper timestamp
-        # interval value) and submit it to the next step.
-        if len(self.__buffers[tick_partition]) > 1:
```

Comment:
I'm pretty confused on what this was supposed to do before. Was this to enforce there being exactly 1 tick per partition?

Reply:
It was supposed to be an optimization to skip checking all the buffers when we know there was an empty one as we don't make progress if any are empty: we only submit the earliest tick to the next stage when we have a tick in the buffer of every partition. So if the current message is added to a non empty buffer then we could infer that one of the other buffers has to be empty and skip the loop below. But.. there is an edge case in which is this not actually true.

Reply:
> But.. there is an edge case in which is this not actually true.

Is this speculative or do we know what this edge case is ?

Reply:
It's not really speculative, if you closely inspect how ticks get removed from the buffer in the loop below you can see how it is possible. Basically it happens when all buffers have got ticks in them except for the one of the current message's partitions AND the message's timestamp is later than some other partition and that other partition's buffer has more than one tick in it already.

Since it's normal for partitions to progress at different speeds it's not even than unusual an edge case and I'm actually surprised we didn't hit the issue before.